### PR TITLE
Tensorflow package integration 

### PR DESCRIPTION
**Pull request by pramodk**
@tristan0x @matz-e  : as you guys are busy with other stuff, I started testing Tensorflow on BB5. I did some fixes to the cherry-picked commits. CPU version is currently building on BB5. Will test GPU version tomorrow.

### CODE REVIEW

**Comment by pramodk** [comment]
```
==> Successfully installed tensorflow
  Fetch: 0.11s.  Build: 26m 12.32s.  Total: 26m 12.43s.
[+] /gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/HPCTM-1206/soft/install/linux-rhel7-x86_64/gcc-6.4.0/tensorflow-1.12.0-xpw5si
```
26min for CPU build.

**Review by matz-e — CHANGES_REQUESTED** [review]
You're using no/outdated license headers.

File: var/spack/repos/builtin/packages/py-astor/package.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+from spack import *
+
+class PyAstor(PythonPackage):
+    """
+    astor is designed to allow easy manipulation of Python source via the AST.
```
Comment (matz-e) [inline var/spack/repos/builtin/packages/py-astor/package.py]: Remove newlines, capitalize.

**Review by matz-e — COMMENTED** [review]
This looks like a pain to figure out/setup. I'm being nitpicky, I don't see any real blocking issues here.

Is this from upstream/will this be upstreamed?

**Comment by pramodk** [comment]
Adding some notes about issues encountered:

* After building cuda support fore more than 1 hour, build failed with : 

```
bazel-out/k8-opt/bin/_solib_local/_U_S_Stensorflow_Scontrib_Srnn_Cgen_Ugru_Uops_Upy_Uwrappers_Ucc___Utensorflow/libtensorflow_framework.so: undefined reference to `cudnnTransformTensor@libcudnn.so.7'
bazel-out/k8-opt/bin/_solib_local/_U_S_Stensorflow_Scontrib_Srnn_Cgen_Ugru_Uops_Upy_Uwrappers_Ucc___Utensorflow/libtensorflow_framework.so: undefined reference to `cudnnSetConvolutionMathType@libcudnn.so.7'
bazel-out/k8-opt/bin/_solib_local/_U_S_Stensorflow_Scontrib_Srnn_Cgen_Ugru_Uops_Upy_Uwrappers_Ucc___Utensorflow/libtensorflow_framework.so: undefined reference to `cudnnPoolingForward@libcudnn.so.7'
bazel-out/k8-opt/bin/_solib_local/_U_S_Stensorflow_Scontrib_Srnn_Cgen_Ugru_Uops_Upy_Uwrappers_Ucc___Utensorflow/libtensorflow_framework.so: undefined reference to `cudnnCreateLRNDescriptor@libcudnn.so.7'
collect2: error: ld returned 1 exit status
Target //tensorflow/tools/pip_package:build_pip_package failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 4132.260s, Critical Path: 863.38s
INFO: 4997 processes: 4997 local.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
==> Error: ProcessError: Command exited with status 1:
    'bazel' 'build' '--jobs=72' '-c' 'opt' '--config=mkl' '--config=cuda' '//tensorflow/tools/pip_package:build_pip_package'

69 errors found in build log:
     16516           ^~~~~~
     16517    ./tensorflow/core/kernels/typed_queue.h:83:7: warning: 'tensorflo
              w::int64 tensorflow::{anonymous}::SizeOf(const SubQueue&) [with S
              ubQueue = std::deque<tensorflow::PersistentTensor>]' defined but
              not used [-Wunused-function]
     16518     int64 SizeOf(const std::deque<PersistentTensor>& sq) {
     16519           ^~~~~~
     16520    ERROR: /gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/HPCTM-1206
              /soft/.stage/kumbhar/spack-stage/spack-stage-oyghjP/tensorflow-1.
              12.0/tensorflow/contrib/rnn/BUILD:257:1: Linking of rule '//tenso
              rflow/contrib/rnn:gen_gru_ops_py_wrappers_cc' failed (Exit 1)
     16521    /usr/bin/ld: warning: libcudnn.so.7, needed by bazel-out/k8-opt/b
              in/_solib_local/_U_S_Stensorflow_Scontrib_Srnn_Cgen_Ugru_Uops_Upy
              _Uwrappers_Ucc___Utensorflow/libtensorflow_framework.so, not foun
              d (try using -rpath or -rpath-link)
  >> 16522    bazel-out/k8-opt/bin/_solib_local/_U_S_Stensorflow_Scontrib_Srnn_
              Cgen_Ugru_Uops_Upy_Uwrappers_Ucc___Utensorflow/libtensorflow_fram
              ework.so: undefined reference to `cudnnCreate@@libcudnn.so.7'
```

Seems like issue with ` --action_env` as discussed [here](https://stackoverflow.com/questions/47080760/tensorflow-fails-to-compile/47295278#47295278). As `cudnn` is not setting `LD_LIBRARY_PATH` and that might be an issue.

* Matthias can explain why following happens with module naming scheme locally : 

```
kumbhar@r2i3n6:/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/HPCTM-1206/spack$ spack module tcl refresh --delete-tree -y
==> Regenerating tcl module files
==> Warning: Could not write module file [/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/HPCTM-1206/soft/modules/tcl/linux-rhel7-x86_64/tensorflow/1.12.0/python3/gpu]
==> Warning: 	--> [Errno 20] Not a directory: u'/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/HPCTM-1206/soft/modules/tcl/linux-rhel7-x86_64/tensorflow/1.12.0/python3/gpu' <--
```

**Comment by pramodk** [comment]
Finally!

```
$ spack install -v --keep-stage --keep-prefix tensorflow+cuda %gcc

==> Warning: Patched overlong shebang in /gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/HPCTM-1206/soft/install/linux-rhel7-x86_64/gcc-6.4.0/tensorflow-1.12.0-7eysmy/bin/saved_model_cli
==> Successfully installed tensorflow
  Fetch: 0.10s.  Build: 1h 21m 52.62s.  Total: 1h 21m 52.72s.
[+] /gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/HPCTM-1206/soft/install/linux-rhel7-x86_64/gcc-6.4.0/tensorflow-1.12.0-7eysmy
```

**Pull request closed without merging**